### ISSUE TITLE
`Callable[[], NoReturn]` does not act like `NoReturn` when called in match statement

### ISSUE THREAD

**Issue by SebbyLaw** [microsoft/pyright#6234]
The following code is not interpreted properly:
```python
from typing import Literal, NoReturn


def raises() -> NoReturn:
    raise RuntimeError("How did you get here?")


def squash(x: Literal[1, 2, 3, 4, 5]) -> int:  # Function with declared return type "int" must return value on all code paths
  Type "None" cannot be assigned to type "int"
    match x:
        case 1 | 2 | 3 | 4:
            return 1
        case 5:
            raises()
```

If I instead, return the result of the `NoReturn` callable
```python
from typing import Literal, NoReturn


def raises() -> NoReturn:
    raise RuntimeError("How did you get here?")


def squash(x: Literal[1, 2, 3, 4, 5]) -> int:  # everything is fine
    match x:
        case 1 | 2 | 3 | 4:
            return 1
        case 5:
            return raises()
```
Or if the call to the `NoReturn` callable is outside of the match statement:
```py
from typing import Literal, NoReturn


def raises() -> NoReturn:
    raise RuntimeError("How did you get here?")


def squash(x: Literal[1, 2, 3, 4, 5]) -> int:  # everything is fine
    match x:
        case 1 | 2 | 3 | 4:
            return 1
        case 5:
            pass

    raises()
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

Reproduce-able on:
- pyright 1.1.333
- ms-python.vscode-pylance-2023.10.40

**Comment by erictraut** [comment]
Thanks for the bug report. The current behavior is a result of a performance optimization I added a long time ago. It affects not only `match` statements but also other similar control-flow constructs like this:

```python
def squash(x: Literal[1, 2]) -> int:
    if x == 1:
        return 1
    elif x == 2:
        raises()
```

I have since mitigated the performance concern in other ways, so I think I can safely remove this performance optimization. This will be included in the next release.
